lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -232,10 +232,35 @@ defmodule Mix.Compilers.Elixir do
 
       sources = update_stale_sources(sources, changed)
       put_compiler_info({modules, structs, sources, pending_modules, %{}})
-      changed
+      {:compile, changed}
     end
   end
 
+  defp dependent_modules(sources, all_modules, pending_modules) do
+    changed_modules =
+      for module(module: module) = entry <- all_modules,
+          entry not in pending_modules,
+          into: %{},
+          do: {module, true}
+
+    Enum.flat_map(pending_modules, fn module(sources: source_files, module: module) ->
+      depending? =
+        Enum.any?(source_files, fn file ->
+          source(
+            compile_references: compile_refs,
```

Comment:
Should this return only runtime dependent modules?

Reply:
Btw, if the code is changed to only check for runtime references the test suite passes. If checking for compile and struct references is needed, I would add a test checking that 🙂

Reply:
@fertapric we need to check compile references because it can be recursive.

Imagine B is changed on disk, so it is compiled. Then A has a runtime dependency on B and C has a compile time dependency on A. Then we should re-verify C. Or maybe we shouldn't? There is nothing that is going to change on A that will lead to a different result in C and the code we verify won't have the compile time dependencies (i.e. they will already have been expanded). So yeah, I think this can keep runtime only.

Reply:
We can use this scenario as an example. Modules `A` and `C` are unchanged, `B` changes the return value of `bar/1`. Will we catch the error?


```elixir
defmodule A do
  defmacro foo() do
    B.bar()
  end
end

defmodule B do
  def bar(), do: :ok # changed to :ok, was :error
end

defmodule C do
  require A

  # will we catch this error?
  def baz(), do: :error = A.foo()
end
```

Reply:
:+1: It has to be recursive on runtime but it can ignore the compile and struct references.